File: cpp/14.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long int

ll chainLength(ll num) {
  ll result = 0;
  while (num != 1) {
    if (num % 2 == 0)
      num /= 2;
    else
      num = 3 * num + 1;

    ++result;
  }

  return result;
}

int main() {
  ll longestLength = INT_MIN, answer;
  for (int i = 1; i <= 1e6; ++i) {
    ll tempChainLength = chainLength(i);
    if (tempChainLength > longestLength) {
      longestLength = tempChainLength;
      answer = i;
    }
  }

  printf("%lld\n", answer);
}

File: py/14.py
def chain_length(num):
    result = 0
    while (num != 1):
        if num % 2 == 0:
            num //= 2
        else:
            num = 3 * num + 1
            
        result += 1
    
    return result

longest_chain_length = -1
answer = 0

for i in range(1, int(1e6 + 1)):
    temp_chain_length = chain_length(i)
    if temp_chain_length > longest_chain_length:
        longest_chain_length = temp_chain_length
        answer = i

print(answer)
